staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go

```diff
@@ -54,18 +53,13 @@ const (
 // FromInt creates an IntOrString object with an int32 value. It is
 // your responsibility not to call this method with a value greater
 // than int32.
-// Deprecated: use FromInt32 instead.
-func FromInt(val int) IntOrString {
-	if val > math.MaxInt32 || val < math.MinInt32 {
-		klog.Errorf("value: %d overflows int32\n%s\n", val, debug.Stack())
-	}
+func FromInt[T constraints.Signed](val T) IntOrString {
 	return IntOrString{Type: Int, IntVal: int32(val)}
```

Comment:
If we keep the stored value as `int32` then we should either force callers to pass `int32` or else keep this range check, shouldn't we?  I know we said "It is your responsibility..." but that is pretty unsatisfying.

So we have several options, most of which you have PR'ed now :)

1) Convert all callers to `FromInt32()` and cast.
2) Convert `FromInt()` to receive `int32` and make all callers cast.
3) Accept all signed ints <= 32 bits, for `int` callers to cast.
4) Change the stored value to int64.

Let's think about deserialization into fields of this type.

If I pass a number which is too big (e.g. 8B) I get an error, either complaioning about "-589934592" (forced cast client-side) or better, an unmarshal error: "json: cannot unmarshal number 8000000000 into Go struct field ServicePort.spec.ports.port of type int32".  Which error you get depends on whether you come in from `update` or `apply`/`patch`.

@apelisse @jpbetz  in case you care about this corner case - basically anywhere we use int32 we have inconsistent deserialization of out of range values.

I know I encouraged you to make this PR increasingly "simple" but I now wonder what's "safe"?  There are a few non-test places that call `intstr.Parse()` with user input, so the stack dump is probably better than silent corruption (considering that I can feed valid int64 values that alias int32.

Even if we convert to int64 (which should be OK because protobuf `varint`) we only push the problem out.  I ACK that someone has to be trying to make trouble to do this (e.g. I can set Deployment replicas to 8 quadrillion, which aliases to a valid int32, but I think that is "GIGO").

So maybe your original PR is actually better?  It doesn't fix deserialization, but it does force explicit callers to at least own their own range. 

#117631 will require callers who take user-input to pre-validate the range

#119285 has some nice updates to `intstr.Parse()`

What do you think?

Reply:
Yes, I prefer approaches which don’t hide problems but push them out to where callers can make sensible decisions about them — since `IntString` (currently) stores an `int32`, let’s make that explicit either through `FromInt32` or changing `FromInt` to only accept an `int32`. Then callers can realise that they’re OK because they’re passing in a port number and it’s already an `int32`, or maybe not because they’ve got a tainted value that hasn’t been sanitised yet.

I do intend to keep the `Parse()` improvements from #119285 in any case — pass in any string and it will be safely stored as an `int32` if it fits, as a `string` otherwise, without the error case of a value passing `Atoi()` but too big for `int32`.

All this raises the question again of what to do about `IntValue()` — for symmetry it should match `FromInt()`. If we want to be explicit (in the vein of the PRs leading to #117631) we should introduce `Int32Value()` and deprecate `IntValue()` in favour of that; if we want to continue down the path of #119285 (which I quite like, even though it means undoing the `FromInt32()` conversions), we should make `IntValue()` return `int32` (which is a breaking change but on an internal API so…).

Reply:
OK.  We've spent too much tim eon the mechanics of this PR, so I am going to say I am convinced that your original approach (explict int32 API, callers need to bounds-check) is OK except for the serialization discussion, which I will recap here.

@skitt do you want to consolidate all the tweaks into that original PR and close the rest?

@apelisse @jpbetz @deads2k @liggitt @smarterclayton @wojtek-t  @msau42 

Long ago, we chose to encode numeric fields in the API as `int32` to be clear about the size (vs `int` which is platform-defined).  Overall this has been fine, except for a few places where we didn't properly bounds-check for negatives (oops) and now zap those to zero and/or issue warnings.  We have a bunch of other fields that are encoded as `int64`, presumably for the same reasons.

The issue comes in with (almost certainly intentional) abuse of API.  E.g. I can set a Deployment's replicas to 8 quadrillion, which aliases to a valid int32, but is clearly wrong.  We don't bounds-check that value.  Basically, as we deserialize from wire-format into int32, we can either lose/corrupt information or we can trigger errors in deserialization.  Those errors are not our usual style of API errors, which is unfortunate.  This is the same reason we transport IP adresses as `string`, so we can deserialize them and produce consistent errors. 

If we had encoded all numeric fields as `int64`, the valid range would be much larger, whihc would allow us more opportunity to catch bad inputs, but doesn't solve for cases like above, where you can legitimately set replicas to values in the billions, and we don't range check. 

Do we care?  If you provide a ridiculously large value that causes int32 or int64 aliasing, are we satisfied to call that "GIGO"?  The alternative would be to convert `int32` -> `int64` or even do more clever deserialization to preserve values.  I think that is not worth the effort, but I wanted to raise it so we all at least ACK the existence of the corner-case.

Lastly, should we encourage/require net new numeric fields to be int64 and to do range checking (e.g. 1M replicas should be enough for anyone) ?

Reply:
> I am going to say I am convinced that your original approach (explict int32 API, callers need to bounds-check) is OK except for the serialization discussion, which I will recap here.

Is the current state of the PR the approach you're wanting to go with? Keeping the unsized int signature marked as Deprecated or removing it entirely seems better than making it silently truncate. The FromInt32 signature seems better to me. Callers can use int32 themselves or cast and bounds-check.

Serialization-wise, int-or-string should already be safe since it encodes as int32 already, right?

> Lastly, should we encourage/require net new numeric fields to be int64 and to do range checking (e.g. 1M replicas should be enough for anyone) ?

Since int64 values > `2^52-1` don't round-trip reliably in json/yaml, I'd keep int32 as the default recommendation for integer fields. If you really need int64, you have to bounds-check values to be within the range that can actually round-trip in all our supported serialization formats.

Reply:
Good point on JSON/YAML brain-damage.  

Will bolster the conventions.

Reply:
@liggitt the original approach Tim is referring to is #116665, which was merged a while back. This discussion started with #117631.

Reply:
> @skitt do you want to consolidate all the tweaks into that original PR and close the rest?

Yes, I can do that. We’ll be left with #117631, #117628, [the “Use intstr.Parse instead of reimplementing it” commit](https://github.com/kubernetes/kubernetes/pull/119285/commits/4023401c8e1ccd95d8355b305e49f30be0e4082b) and [the `Parse` tweaks](https://github.com/kubernetes/kubernetes/pull/119285/files#diff-4e889356ab2649590590aaae2ac75f32e7b1cc8c4576a01304c120ea7da71365R65) from  from #119285 (in new PRs).